platform/ios/src/MGLMapView+OpenGL.mm

```diff
@@ -0,0 +1,251 @@
+    resource.glView.contentScaleFactor = [UIScreen instancesRespondToSelector:@selector(nativeScale)]
+                                    ? [[UIScreen mainScreen] nativeScale]
+                                    : [[UIScreen mainScreen] scale];
+    resource.glView.contentMode = UIViewContentModeCenter;
+    resource.glView.drawableStencilFormat = GLKViewDrawableStencilFormat8;
+    resource.glView.drawableDepthFormat = GLKViewDrawableDepthFormat16;
+    resource.glView.opaque = mapView.opaque;
+    resource.glView.layer.opaque = mapView.opaque;
+    resource.glView.enableSetNeedsDisplay = NO;
+    CAEAGLLayer* eaglLayer = MGL_OBJC_DYNAMIC_CAST(resource.glView.layer, CAEAGLLayer);
+    eaglLayer.presentsWithTransaction = NO;
+
+    [mapView insertSubview:resource.glView atIndex:0];
+    size = resource.framebufferSize();
+}
+
+UIView* MGLMapViewOpenGLImpl::getView() {
+    auto& resource = getResource<MGLMapViewOpenGLRenderableResource>();
+    return resource.glView;
+}
+
+void MGLMapViewOpenGLImpl::deleteView() {
+    auto& resource = getResource<MGLMapViewOpenGLRenderableResource>();
+    [resource.glView deleteDrawable];
+}
+
+#ifdef MGL_RECREATE_GL_IN_AN_EMERGENCY
+// See https://github.com/mapbox/mapbox-gl-native/issues/14232
+void MGLMapViewOpenGLImpl::emergencyRecreateGL() {
+    auto& resource = getResource<MGLMapViewOpenGLRenderableResource>();
+    MGLLogError(@"Rendering took too long - creating GL views");
+
+    CAEAGLLayer* eaglLayer = MGL_OBJC_DYNAMIC_CAST(resource.glView.layer, CAEAGLLayer);
+    eaglLayer.presentsWithTransaction = NO;
+
+    [mapView pauseRendering:nil];
+
+    // Just performing a pauseRendering:/resumeRendering: pair isn't sufficient - in this case
+    // we can still get errors when calling bindDrawable. Here we completely
+    // recreate the GLKView
+
+    [mapView.userLocationAnnotationView removeFromSuperview];
+    [resource.glView removeFromSuperview];
+
+    // Recreate the view
+    resource.glView = nil;
+    createView();
+
+    if (mapView.annotationContainerView) {
+        [resource.glView insertSubview:mapView.annotationContainerView atIndex:0];
+    }
+
+    [mapView updateUserLocationAnnotationView];
+
+    // Do not bind...yet
+
+    if (mapView.window) {
+        [mapView resumeRendering:nil];
+        eaglLayer = MGL_OBJC_DYNAMIC_CAST(resource.glView.layer, CAEAGLLayer);
+        eaglLayer.presentsWithTransaction = mapView.enablePresentsWithTransaction;
+    } else {
+        MGLLogDebug(@"No window - skipping resumeRendering");
+    }
+}
+#endif
+
+mbgl::gl::ProcAddress MGLMapViewOpenGLImpl::getExtensionFunctionPointer(const char* name) {
+    static CFBundleRef framework = CFBundleGetBundleWithIdentifier(CFSTR("com.apple.opengles"));
+    if (!framework) {
+        throw std::runtime_error("Failed to load OpenGL framework.");
+    }
+
+    CFStringRef str = CFStringCreateWithCString(kCFAllocatorDefault, name, kCFStringEncodingASCII);
```

Comment:
Nit: Use NSString instead of CFString for clarity and less memory management risk.

Reply:
What's the risk? I didn't modify this code, only moved it over from MGLMapView.mm

Reply:
Some (if not all) of Core Foundation’s API is not covered by automatic reference counting so you'd to manually have to call `CFRelease()` as seen below. This method could probably be rewritten to use NSBundle instead of CFBundle and NSString instead of CFStringRef but that might be out of scope for this PR.

Reply:
I don't think there's a way to get function pointers from a `NSBundle` object, but I rewrote it to use `NSString` with a `__bridge` call to `CFStringRef`.